python/GafferSceneUI/ShaderQueryUI.py

```diff
@@ -0,0 +1,426 @@
+			Gaffer.IntPlug,
+			"NumericDivider",
+			Gaffer.StringPlug,
+			"StringDivider",
+			Gaffer.V2iPlug,
+			Gaffer.V3iPlug,
+			Gaffer.V2fPlug,
+			Gaffer.V3fPlug,
+			"VectorDivider",
+			Gaffer.Color3fPlug,
+			Gaffer.Color4fPlug,
+		] :
+			if isinstance( item, six.string_types ) :
+				result.append( "/" + item, { "divider": True } )
+			else :
+				result.append(
+					"/" + item.__name__.replace( "Plug", "" ),
+					{
+						"command" : functools.partial( Gaffer.WeakMethod( self.__addQuery ), "parameter", item ),
+					}
+				)
+
+		return result
+
+	def __addFromLocationMenuDefinition( self ) :
+
+		with self.getContext() :
+			node = GafferSceneUI.ShaderUI._shaderSourceNode( self, GafferScene.ShaderQuery )
+
+			if node is None :
+				return {}
+
+			useFullAttributes = node["inherit"].getValue()
+			attributeNamePatterns = node["shader"].getValue()
+			location = node["location"].getValue()
+
+			return GafferSceneUI.ShaderUI._shaderParameterMenuDefinition(
+				node["scene"],
+				[location],
+				self.__addQuery,
+				attributeNamePatterns,
+				useFullAttributes
+			)
+
+	def __addQuery( self, name, plugTypeOrValue ) :
+
+		with Gaffer.UndoScope( self.getPlug().ancestor( Gaffer.ScriptNode ) ) :
+
+			node = self.getPlug().node()
+
+			if isinstance( plugTypeOrValue, IECore.Data ) :
+				dummyPlug = Gaffer.PlugAlgo.createPlugFromData(
+					"dummyPlug",
+					Gaffer.Plug.Direction.In,
+					Gaffer.Plug.Flags.Default,
+					plugTypeOrValue
+				)
+				node.addQuery( dummyPlug, name)
+			else:
+				node.addQuery( plugTypeOrValue(), name)
+
+	def __updateQueryMetadata( self, plug ) :
+
+		node = plug.node()
+
+		if node["queries"].isAncestorOf( plug ) :
+
+			idx = node.getChildIndex( node["queries"], plug )
+
+			if node["queries"][idx]["name"] == plug :
+
+				Gaffer.Metadata.plugValueChangedSignal( node )(
+					node["exists"][idx],
+					"label",
+					Gaffer.Metadata.ValueChangedReason.StaticRegistration
+				)
+
+				Gaffer.Metadata.plugValueChangedSignal( node )(
+					node["values"][idx],
+					"label",
+					Gaffer.Metadata.ValueChangedReason.StaticRegistration
+				)
+
+##########################################################################
+# Delete Plug
+##########################################################################
+
+
+def __plugPopupMenu( menuDefinition, plugValueWidget ) :
+
+	readOnlyUI = plugValueWidget.getReadOnly()
+	plug = plugValueWidget.getPlug().ancestor( Gaffer.NameValuePlug )
+
+	if plug is not None :
+
+		if len( menuDefinition.items() ) :
+			menuDefinition.append( "/DeleteDivider", { "divider" : True } )
+
+		menuDefinition.append( "/Delete", { "command" : functools.partial( __deletePlug, plug ), "active" : not readOnlyUI and not Gaffer.MetadataAlgo.readOnly( plug ) } )
+
+def __deletePlug( plug ) :
+
+	with Gaffer.UndoScope( plug.ancestor( Gaffer.ScriptNode ) ) :
+		plug.node().removeQuery( plug )
+
+GafferUI.PlugValueWidget.popupMenuSignal().connect( __plugPopupMenu, scoped = False )
```

Comment:
I tried that first, but realized it messes up the removal of corresponding outputs for the query. I tried connecting a `ChildRemovedSignal` but it seems that by the time the signal is firing, the query plug is already out of the children list, preventing us from getting the index.

If that's correct, it may be another argument for abandoning the indexed scheme and using plug names to link query to output.

Reply:
Ah, of course. Sorry, I should have realised. Thanks for explaining...